In [ ]:
# Installation automatique des dépendances requises dans le noyau Jupyter actuel
%pip install -r ../requirements.txt

# 🧠 Étape 5 : Modélisation (Machine Learning & Deep Learning) (Squelette Étudiant)

Cette étape correspond au cinquième chapitre du cours. L'objectif est d'implémenter d'une part un modèle de Machine Learning tabulaire (ex: RandomForest) et d'autre part un réseau de neurones convolutif (CNN) sous TensorFlow pour traiter des images ou signaux complexes.

### 1. Préparation de l'environnement

In [ ]:
import os
import pandas as pd
import numpy as np

from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split

print("Librairies de modélisation (Machine Learning Classique) importées avec succès !")

Librairies de modélisation (Machine Learning Classique) importées avec succès !


### 🤖 Modélisation : Prédiction de la consommation (MPG)

**Le "Pourquoi" de notre algorithme :** L'objectif est de prédire une valeur numérique continue (la consommation `mpg`), il s'agit donc d'un problème de **Régression**. 
Lors de notre Analyse Exploratoire (EDA), nous avons constaté que les relations entre le poids, la puissance et la consommation formaient des courbes et non des lignes droites parfaites. C'est pourquoi nous avons écarté la simple Régression Linéaire au profit d'un **Random Forest Regressor** (Forêt Aléatoire). Cet algorithme est robuste face aux valeurs extrêmes (*outliers*) que nous avons conservées et gère parfaitement les relations non-linéaires.

**Prévention de la triche (Data Leakage) :** Avant d'entraîner le modèle, nous devons retirer la colonne `mpg_category` que nous avions créée lors du Feature Engineering. Si nous laissions cette colonne à l'algorithme, il "tricherait" en connaissant déjà la catégorie de consommation pour prédire le chiffre exact !

In [15]:
# 1. Chargement des données propres (issues du Wrangling et Feature Engineering)
df = pd.read_csv('../data/processed/cleaned_data_sample.csv')

# 2. Définition de la cible (y) et des caractéristiques (X)
y = df['mpg']

# On supprime la cible (mpg) ET la colonne catégorie (si elle existe) pour éviter la triche
X = df.drop(columns=['mpg', 'mpg_category'], errors='ignore')

# 3. Découpage pour éviter le surapprentissage : 80% Entraînement / 20% Test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(f"Véhicules pour l'entraînement : {X_train.shape[0]}")
print(f"Véhicules pour le test (gardés secrets) : {X_test.shape[0]}")

# 4. Instanciation et Entraînement du modèle
rf_model = RandomForestRegressor(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)

print("\n✅ Modèle Random Forest entraîné avec succès !")

Véhicules pour l'entraînement : 318
Véhicules pour le test (gardés secrets) : 80

✅ Modèle Random Forest entraîné avec succès !


### 📏 1. Évaluation des Performances (Les Métriques)

**Le "Pourquoi" de ces métriques :** Pour savoir si notre modèle est bon, nous ne pouvons pas juste regarder les chiffres à l'œil nu. Nous utilisons trois métriques mathématiques standards de la régression :
* **Le MAE (Erreur Absolue Moyenne) :** Il nous dit, en moyenne, de combien de MPG notre modèle se trompe. C'est la métrique la plus facile à comprendre pour un être humain.
* **Le RMSE (Racine de l'Erreur Quadratique Moyenne) :** Il pénalise très fortement les grosses erreurs. Si notre modèle se trompe énormément sur une seule voiture, le RMSE va exploser.
* **Le R² (Coefficient de Détermination) :** C'est un pourcentage (de 0 à 1). Il nous dit quel pourcentage de la consommation est réellement expliqué par notre modèle (grâce au poids, la puissance, etc.). Plus il est proche de 1, plus le modèle est un "devin" parfait.

In [24]:
# On utilise les prédictions du Random Forest
y_pred = y_pred_rf

# Calcul des métriques
mae = mean_absolute_error(y_test, y_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)

print(f"Erreur Absolue Moyenne (MAE)  : {mae:.2f} MPG")
print(f"Racine de l'Erreur Quadratique (RMSE) : {rmse:.2f} MPG")
print(f"Score R² (Précision globale)  : {r2:.2f}")

Erreur Absolue Moyenne (MAE)  : 1.61 MPG
Racine de l'Erreur Quadratique (RMSE) : 2.15 MPG
Score R² (Précision globale)  : 0.91


#### Interprétation des résultats

| Métrique | Valeur | Signification |
|---|---|---|
| **MAE** | 1.61 MPG | En moyenne, notre modèle se trompe de 1.61 miles par gallon |
| **RMSE** | 2.15 MPG | Les grosses erreurs sont rares — le modèle est stable |
| **R²** | 0.91 | Le modèle explique **91% de la variance** du MPG |

**Ce que ça veut dire en clair :**

- Un R² de 0.91 est **excellent** — le modèle capture très bien les patterns
- Une MAE de 1.61 MPG signifie que si une voiture fait 25 MPG réellement,
  notre modèle prédit entre 23.4 et 26.6 MPG en moyenne
- Le Random Forest fait clairement mieux qu'une simple régression linéaire,
  ce qui confirme que les relations entre les variables **ne sont pas linéaires**

### 2. Modélisation Tabulaire (Machine Learning)

**À COMPLÉTER PAR L'ÉTUDIANT :**
Entraînez un modèle d'apprentissage supervisé (ex: forêt aléatoire) sur les caractéristiques extraites de votre jeu de données.

In [ ]:
# Chargement des données propres et ingénierie des caractéristiques
df = pd.read_csv('../data/processed/cleaned_data_sample.csv')
df['timestamp'] = pd.to_datetime(df['timestamp'])
df_feat = dc.feature_engineering(df, 'timestamp')

features = ['hour', 'dayofweek']
target = 'value'

# Split chronologique simple pour l'entraînement
X_train = df_feat[features].iloc[:-4]
y_train = df_feat[target].iloc[:-4]

# TODO: Instancier et entraîner RandomForestRegressor sur (X_train, y_train)
rf_model = RandomForestRegressor(n_estimators=10, random_state=42)
rf_model.fit(X_train, y_train)
print("Modèle de forêt aléatoire entraîné !")

### 3. Modélisation Vision / Deep Learning (CNN & TensorFlow)

**À COMPLÉTER PAR L'ÉTUDIANT :**
Pour des motifs complexes (images/signaux), mettez en place un réseau convolutif (Conv2D, Pooling, Dense) pour classifier ou enrichir vos prédictions.

In [ ]:
# Génération fictive d'un jeu d'images simples (64x64 pixels) de cercles (Classe 0) vs rectangles (Classe 1)
def generate_dummy_images(num_samples=100):
    images = np.zeros((num_samples, 64, 64, 3), dtype=np.float32)
    labels = np.zeros(num_samples, dtype=np.int32)
    for i in range(num_samples):
        label = np.random.choice([0, 1])
        labels[i] = label
        images[i, :, :, :] = 0.2 + np.random.normal(0, 0.01, (64, 64, 3))
        if label == 1:
            images[i, 10:30, 10:30, 0] = 0.8
        else:
            images[i, 20:40, 20:40, 1] = 0.8
    return images, labels

X_images, y_labels = generate_dummy_images(100)
split = int(0.8 * len(X_images))
X_img_train = X_images[:split]
y_img_train = y_labels[:split]

print(f"Dataset d'images brutes généré. Dimensions Train : {X_img_train.shape}")

In [ ]:
# TODO: Définir l'architecture séquentielle du CNN avec layers.Conv2D et layers.MaxPooling2D
cnn_model = models.Sequential([
    layers.Conv2D(16, (3, 3), activation='relu', input_shape=(64, 64, 3)),
    layers.MaxPooling2D((2, 2)),
    layers.Flatten(),
    layers.Dense(16, activation='relu'),
    layers.Dense(1, activation='sigmoid')
])

cnn_model.summary()

In [ ]:
# TODO: Compiler et entraîner le CNN avec l'optimiseur adam et une binary_crossentropy
cnn_model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
cnn_model.fit(X_img_train, y_img_train, epochs=2, batch_size=32, verbose=1)
print("CNN entraîné avec succès !")